src/clean-core/detail/vector_base.hh

```diff
@@ -167,23 +172,54 @@ public:
     void push_back_range(Range&& range)
     {
         static_assert(cc::is_any_range<Range>);
+        static_assert(std::is_copy_constructible_v<T>, "only works with copyable types. use push_back(T&&) to move elements into the vector");
 
-        if constexpr (collection_traits<Range>::has_size)
-            this->reserve(_size + cc::collection_size(range));
-
-        if constexpr (cc::is_contiguous_range<Range, T> && std::is_trivially_copyable_v<T> && std::is_trivially_destructible_v<T>)
+        if constexpr (cc::is_contiguous_range<Range, T>)
         {
-            std::memcpy(&_data[_size], range.data(), range.size() * sizeof(T));
-            _size += range.size();
+            size_t const num_new_elems = cc::collection_size(range);
+            if (num_new_elems == 0)
+            {
+                return;
+            }
+
+            this->reserve(_size + num_new_elems);
+
+            // NOTE: this would assert for an empty span<T>
+            // only get this pointer after checking for size == 0
+            T const* const new_elem_data = &range[0];
+            detail::container_copy_construct_range<T>(new_elem_data, num_new_elems, &_data[_size]);
+            _size += num_new_elems;
         }
         else
         {
-            // TODO: a few optimizations are possible here (like moving or in-place construction)
+            if constexpr (collection_traits<Range>::has_size)
+                this->reserve(_size + cc::collection_size(range));
+
             for (auto&& v : range)
                 this->push_back(v);
         }
     }
```

Comment:
Note: broken for this vector itself as arg or spans/subspans into it, also potentially fix for const ref to array